### PR TITLE
WIP: user-cluster-controller part1

### PR DESCRIPTION
**Pull request by thz**
**What this PR does / why we need it**:

See `docs/proposals/user-cluster-controller.md`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
Introduces the user-cluster-controller.
```

### CODE REVIEW

File: api/pkg/controller/cluster/launching.go [hunk 1]
```diff
@@ -63,7 +63,19 @@ func (cc *Controller) launchingCreateClusterInfoConfigMap(c *kubermaticv1.Cluste
...
+				"clusterName": c.Name,
+			}
+			var (
+				vpnAddress string
+				errVpnAddr error
+			)
+			if vpnAddress, errVpnAddr = cc.getOpenVpnAddress(c); errVpnAddr != nil {
+				return fmt.Errorf("failed to determine master vpn address: %v", err)
+			}
+			cm.Data["masterVpnAddress"] = vpnAddress
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/launching.go:61]: The cluster-info ConfigMap which is defined by the kubeadm project, i would rather avoid using it for our own data.
I would prefer to pass the cluster-name and the vpn-address as flag to the user-cluster-resources-controller-manager

File: api/pkg/controller/usercluster/resources.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	creators := GetUserClusterRoleCreators(data)
+
+	for _, create := range creators {
+		var existing *rbacv1.ClusterRole
+		cRole, err := create(data, nil)
+		if err != nil {
+			return fmt.Errorf("failed to build ClusterRole: %v", err)
+		}
+
+		if existing, err = ucc.client.RbacV1().ClusterRoles().Get(cRole.Name, metav1.GetOptions{}); err != nil {
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go:76]: Use a lister

File: api/pkg/controller/usercluster/resources.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+			if !errors.IsNotFound(err) {
+				return err
+			}
+
+			if _, err = ucc.client.RbacV1().ClusterRoles().Create(cRole); err != nil {
+				return fmt.Errorf("failed to create ClusterRole %s: %v", cRole.Name, err)
+			}
+			glog.V(4).Infof("Created ClusterRole %s inside user-cluster %s", cRole.Name, data.ClusterNameOrEmpty())
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go]: Remove ` inside user-cluster %s` - as its obvios

File: api/pkg/controller/usercluster/resources.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	for _, create := range creators {
+		var existing *corev1.ConfigMap
+		cm, err := create(data, nil)
+		if err != nil {
+			return fmt.Errorf("failed to build ConfigMap: %v", err)
+		}
+
+		if existing, err = ucc.client.CoreV1().ConfigMaps(cm.Namespace).Get(cm.Name, metav1.GetOptions{}); err != nil {
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go:125]: Use a lister

File: api/pkg/controller/usercluster/resources.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+			if !errors.IsNotFound(err) {
+				return err
+			}
+
+			if _, err = ucc.client.CoreV1().ConfigMaps(cm.Namespace).Create(cm); err != nil {
+				return fmt.Errorf("failed to create ConfigMap %s: %v", cm.Name, err)
+			}
+			glog.V(4).Infof("Created ConfigMap %s inside user-cluster %s", cm.Name, data.ClusterNameOrEmpty())
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go]: Remove ` inside user-cluster %s` as its obvios

File: api/pkg/controller/usercluster/resources.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		if equality.Semantic.DeepEqual(cRole, existing) {
+			continue
+		}
+
+		if _, err = ucc.client.RbacV1().ClusterRoles().Update(cRole); err != nil {
+			return fmt.Errorf("failed to update ClusterRole %s: %v", cRole.Name, err)
+		}
+		glog.V(4).Infof("Updated ClusterRole %s inside user-cluster %s", cRole.Name, data.ClusterNameOrEmpty())
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go]: remove ` inside user-cluster %s` as its obvios

File: api/pkg/controller/usercluster/resources.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		if equality.Semantic.DeepEqual(cm, existing) {
+			continue
+		}
+
+		if _, err = ucc.client.CoreV1().ConfigMaps(cm.Namespace).Update(cm); err != nil {
+			return fmt.Errorf("failed to update ConfigMap %s: %v", cm.Name, err)
+		}
+		glog.V(4).Infof("Updated ConfigMap %s inside user-cluster %s", cm.Name, data.ClusterNameOrEmpty())
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go]: Remove inside user-cluster %s as its obvios

File: api/pkg/controller/usercluster/usercluster_controller.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+	// Get confimaps from lister, make a copy.
+	cachedConfigMaps, err := ucc.configMapLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to receive configMaps from lister: %v", err)
+	}
+	configMaps := make([]*corev1.ConfigMap, len(cachedConfigMaps))
+	for i := range cachedConfigMaps {
+		configMaps[i] = cachedConfigMaps[i].DeepCopy()
+	}
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/usercluster_controller.go]: What is this for?

File: api/pkg/resources/openvpn/configmap.go [hunk 9]
```diff
@@ -84,17 +82,22 @@ func ClientConfigConfigMap(data *resources.TemplateData, existing *corev1.Config
...
 
-	config := fmt.Sprintf(`client
+func getOpenVPNConfig(remoteHost, remotePort string) string {
+	return fmt.Sprintf(`client
 proto tcp
 dev kube
 dev-type tun
 auth-nocache
-remote %s %d
+remote %s %s
```
Comment (mrIncompetent) [inline api/pkg/resources/openvpn/configmap.go:98]: Why changing to `%s` when we have a port?

File: api/pkg/resources/resources.go [hunk 10]
```diff
@@ -138,6 +140,8 @@ const (
 	ControllerManagerCertUsername = "system:kube-controller-manager"
 	//SchedulerCertUsername is the name of the user coming from kubeconfig cert
 	SchedulerCertUsername = "system:kube-scheduler"
+	//UserClusterControllerManagerCertUsername is the name of the user coming from kubeconfig cert
+	UserClusterControllerManagerCertUsername = "kubermatic:user-cluster-controller-manager"
```
Comment (mrIncompetent) [inline api/pkg/resources/resources.go:144]: Woulnd't it be easier if we just use the `system:masters` group here?
So we don't run into the chicken-egg problem, where we need to create rbac rules inside the cluster to make the uccm work?

File: api/pkg/resources/resources.go [hunk 11]
```diff
@@ -265,6 +271,10 @@ const (
 // ConfigMapCreator defines an interface to create/update ConfigMap's
 type ConfigMapCreator = func(data *TemplateData, existing *corev1.ConfigMap) (*corev1.ConfigMap, error)
 
+// UserConfigMapCreator is the same as ConfigMapCreator but with a UserClusterData input.
+// The two should definitely be harmonized (interface perhaps).
+type UserConfigMapCreator = func(data *UserClusterData, existing *corev1.ConfigMap) (*corev1.ConfigMap, error)
```
Comment (mrIncompetent) [inline api/pkg/resources/resources.go]: Shouldn't it be `UserClusterConfigMapCreator` ?

File: api/pkg/resources/resources.go [hunk 12]
```diff
@@ -283,6 +293,9 @@ type RoleBindingCreator = func(data *TemplateData, existing *rbacv1.RoleBinding)
 // ClusterRoleCreator defines an interface to create/update RBAC ClusterRoles
 type ClusterRoleCreator = func(data *TemplateData, existing *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error)
 
+// UserClusterRoleCreator defines an interface to create/update RBAC ClusterRoles
+type UserClusterRoleCreator = func(data *UserClusterData, existing *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error)
```
Comment (mrIncompetent) [inline api/pkg/resources/resources.go]: Group all user-cluster interfaces

File: api/pkg/resources/userclustercontrollermanager/deployment.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	dep.Spec.Template.Spec.Containers = []corev1.Container{
+		{
+			Name:            name,
+			Image:           data.ImageRegistry(resources.RegistryQuay) + "/kubermatic/user-cluster-controller-manager:v0.1.0-dev10",
+			ImagePullPolicy: corev1.PullIfNotPresent,
+			Command:         []string{"/usr/local/bin/user-cluster-controller-manager"},
+			Args: []string{
+				"--kubeconfig", "/etc/kubernetes/kubeconfig/kubeconfig",
+				"--logtostderr", "--v", "6",
```
Comment (mrIncompetent) [inline api/pkg/resources/userclustercontrollermanager/deployment.go]: Should be turned down to 4 before merging

Comment (mrIncompetent) [inline api/pkg/resources/userclustercontrollermanager/deployment.go]:
```go
"--logtostderr", 
"--v", "6",
```

Comment (kdomanski) [reply api/pkg/resources/userclustercontrollermanager/deployment.go]: Why do we need to run everything in extra verbose mode?

Comment (mrIncompetent) [reply api/pkg/resources/userclustercontrollermanager/deployment.go]: See https://github.com/kubermatic/kubermatic/pull/1956#discussion_r215564316

Comment (kdomanski) [reply api/pkg/resources/userclustercontrollermanager/deployment.go]: My bad.

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (thz) [reply api/pkg/controller/cluster/launching.go:61]: Created a dedicated configmap "cluster-seed" now.

File: api/pkg/controller/usercluster/usercluster_controller.go (hunk 8 above)
Comment (thz) [reply api/pkg/controller/usercluster/usercluster_controller.go]: Ancient relic. It's gone.

File: api/pkg/resources/openvpn/configmap.go (hunk 9 above)
Comment (thz) [reply api/pkg/resources/openvpn/configmap.go:98]: The actual address as a host, port pair is retrieved from the `cluster-seed` configmap. There I used the more common notation of `host:port` for the "address" of the vpn endpoint. I see no added value in separating the two nor do I see value in handing the port as an integer. So in doubt I prefer less code, less fields, less conversion.

Objections?

(Previously in the cluster-controller this was indeed an integer because it was fetched as `openvpnSvc.Spec.Ports[0].NodePort`)

File: api/pkg/resources/resources.go (hunk 10 above)
Comment (thz) [reply api/pkg/resources/resources.go:144]: This is not the classical chicken-egg problem. It lacks the back-dependency. We need to create a deployment in the cluster-controller, so we can add the RBAC as well.
Isn't your argument here more about removing complexity by sacrificing explicit RBAC?

That said, I do not have a strong opinion, but wouldn't we get a better audit trail in the end if the uccm uses a dedicated (named) role?

File: api/cmd/user-cluster-controller-manager/controllers.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+	return g.Run()
+}
+
+func createUserClusterController(ctrlCtx *controllerContext) (runner, error) {
+
+	return usercluster.NewController(
+		ctrlCtx.kubeClient,
+		ctrlCtx.kubeInformerFactory.Core().V1().ConfigMaps(),
+		ctrlCtx.kubeInformerFactory.Core().V1().Services())
+}
```
Comment (mrIncompetent) [inline api/cmd/user-cluster-controller-manager/controllers.go:73]: Rename everything inside this file from `UserClusterController` to `Controller`.

File: api/cmd/user-cluster-controller-manager/main.go [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+	flag.StringVar(&runOp.kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&runOp.masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&runOp.internalAddr, "internal-address", "127.0.0.1:8085", "The address on which the internal ser    ver is running on")
+	flag.StringVar(&runOp.workerName, "worker-name", "", "The name of the worker that will only processes resources with label=worker-name.")
+	flag.IntVar(&runOp.workerCount, "worker-count", 4, "Number of workers which process the clusters in parallel.")
+	flag.StringVar(&runOp.overwriteRegistry, "overwrite-registry", "", "registry to use for all images")
+	flag.StringVar(&runOp.dockerPullConfigJSONFile, "docker-pull-config-json-file", "config.json", "The file containing the docker auth config.")
+	flag.Parse()
+
+	if runOp.dockerPullConfigJSONFile == "" {
```
Comment (mrIncompetent) [inline api/cmd/user-cluster-controller-manager/main.go]: What for do we need this here?
We don't run any private images inside customer clusters as far as i know

File: api/cmd/user-cluster-controller-manager/main.go [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+	}
+	glog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.V(4).Infof)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: masterKubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: controllerName})
+	return recorder, nil
+}
+
+func newUserClusterControllerContext(runOp controllerRunOptions, done <-chan struct{}, kubeClient kubernetes.Interface) *controllerContext {
```
Comment (mrIncompetent) [inline api/cmd/user-cluster-controller-manager/main.go]: Rename to `newControllerContext`

File: api/cmd/user-cluster-controller-manager/main.go [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+	//Register the global error metric. Ensures that runtime.HandleError() increases the error metric
+	metrics.RegisterRuntimErrorMetricCounter("user_cluster_controller_manager", prometheus.DefaultRegisterer)
+
+	stopCh := signals.SetupSignalHandler()
+	ctx, ctxDone := context.WithCancel(context.Background())
+
+	// Create Context
+	ctrlCtx := newUserClusterControllerContext(runOp, ctx.Done(), kubeClient)
+
+	controllers, err := createAllUserClusterControllers(ctrlCtx)
```
Comment (mrIncompetent) [inline api/cmd/user-cluster-controller-manager/main.go]: rename to `createAllControllers`

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/controller/cluster/launching.go:61]: Can we find a more precise name her? Maybe even `cluster-info-extras`
`cluster-seed` seems a bit non-intuitive

File: api/pkg/controller/cluster/launching.go [hunk 18]
```diff
@@ -128,6 +168,17 @@ func (cc *Controller) launchingCreateClusterInfoConfigMap(c *kubermaticv1.Cluste
 	return nil
 }
 
+func (cc *Controller) getOpenVpnAddress(c *kubermaticv1.Cluster) (string, error) {
+	s, err := cc.serviceLister.Services(c.Status.NamespaceName).Get(resources.OpenVPNServerServiceName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.V(2).Infof("NotFound in getOpenVpnAddress is unexpected: %v", err)
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/launching.go:177]: Whats the benefit of logging it here if we return the error anyway on https://github.com/kubermatic/kubermatic/pull/1956/files#diff-a87fac085c1ef1d73a261323902c708eR177 ?

All errors are being bubbled up to the controller-error handler & logged there.

File: api/pkg/controller/rbac/sync_project.go [hunk 19]
```diff
@@ -192,7 +192,7 @@ func ensureClusterRBACRoleForResource(kubeClient kubernetes.Interface, groupName
 		return err
 	}
 
-	_, err = kubeClient.RbacV1().ClusterRoles().Create(generatedClusterRole)
+	_, err = kubeClient.RbacV1().ClusterRoles().Create(generatedClusterRole) // ownerref missing FIXME
```
Comment (mrIncompetent) [inline api/pkg/controller/rbac/sync_project.go:214]: What ownerRef are we missing here?

File: api/pkg/controller/usercluster/resources.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+
+	// get initial data from seed - this verifies the availability of seed data
+	name, err := data.GetClusterName()
+	if err != nil {
+		return fmt.Errorf("failed to get user-cluster name: %v", err)
+	}
+
+	if len(name) == 0 {
+		return fmt.Errorf("empty user-cluster name")
+	}
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/resources.go]: Move this error handling to `data.GetClusterName()` as it already returns an error

File: api/pkg/controller/usercluster/usercluster_controller.go [hunk 21]
```diff
@@ -0,0 +1,121 @@
+package usercluster
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+
+	// kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+	// "github.com/kubermatic/kubermatic/api/pkg/resources"
```
Comment (mrIncompetent) [inline api/pkg/controller/usercluster/usercluster_controller.go:9]: Can those be removed?

File: api/pkg/resources/data.go [hunk 22]
```diff
@@ -429,3 +436,43 @@ func (d *UserClusterData) InClusterPrometheusDisableDefaultScrapingConfigs() boo
...
+	cm, err := d.ConfigMapLister().ConfigMaps(metav1.NamespacePublic).Get(ClusterSeedConfigMapName)
+	if err != nil {
+		return "", fmt.Errorf("could not get configmap %s: %v", ClusterSeedConfigMapName, err)
+	}
+
+	return cm.Data["clusterName"], nil
+}
+
+// ClusterNameOrEmpty returns the name of the cluster or empty string in case of error
+func (d *UserClusterData) ClusterNameOrEmpty() string {
```
Comment (mrIncompetent) [inline api/pkg/resources/data.go]: Why do we need this function?

File: api/pkg/resources/data.go [hunk 23]
```diff
@@ -429,3 +436,43 @@ func (d *UserClusterData) InClusterPrometheusDisableDefaultScrapingConfigs() boo
...
+	n, e := d.GetClusterName()
+	if e != nil {
+		return ""
```
Comment (mrIncompetent) [inline api/pkg/resources/data.go]: At least log the error.

Comment (mrIncompetent) [reply api/pkg/resources/data.go]: Better: use `runtime.HandleErr`

File: api/pkg/resources/data.go [hunk 24]
```diff
@@ -429,3 +436,43 @@ func (d *UserClusterData) InClusterPrometheusDisableDefaultScrapingConfigs() boo
...
+	}
+	return n
+}
+
+// IpamEnabled returns true iff ipam shall happen
+func (d *UserClusterData) IpamEnabled() bool {
+	// this should be evaluated by checking for len(Spec.MachineNetworks)>0
+	return false
```
Comment (mrIncompetent) [inline api/pkg/resources/data.go]: We should not break existing functionality.
Must be fixed before we can merge this

File: api/pkg/resources/data.go [hunk 25]
```diff
@@ -429,3 +436,43 @@ func (d *UserClusterData) InClusterPrometheusDisableDefaultScrapingConfigs() boo
...
+func (d *UserClusterData) GetMasterVpnAddress() (string, string, error) {
+	cm, err := d.ConfigMapLister().ConfigMaps(metav1.NamespacePublic).Get(ClusterSeedConfigMapName)
+	if err != nil {
+		return "", "", fmt.Errorf("could not get configmap %s: %v", ClusterSeedConfigMapName, err)
+	}
+
+	var vpnHost, vpnPort string
+	if idx := strings.LastIndex(cm.Data["masterVpnAddress"], ":"); idx > -1 {
+		vpnHost = cm.Data["masterVpnAddress"][:idx]
+		vpnPort = cm.Data["masterVpnAddress"][idx+1:]
```
Comment (mrIncompetent) [inline api/pkg/resources/data.go:455]: Add validation for the `vpnPort` by casting it to int.
Otherwise we accept `:` as valid address.

Comment (mrIncompetent) [reply api/pkg/resources/data.go:455]: Also then return the int value, not the string for the port

File: api/pkg/resources/openvpn/configmap.go (hunk 9 above)
Comment (mrIncompetent) [reply api/pkg/resources/openvpn/configmap.go:98]: Yes, see https://github.com/kubermatic/kubermatic/pull/1956/files#r216936449

If we cast to int, we fail early in case we have invalid data instead of silently ignoring it

File: api/pkg/resources/resources.go (hunk 10 above)
Comment (mrIncompetent) [reply api/pkg/resources/resources.go:144]: You're right, i was assuming we move all user-cluster resource handling into the uccm - which would create a chicken-egg-problem.

We're not loosing anything for the audit trail, if we use a dedicated username + the `system:masters` group.
All actions would be logged with the username + group anyway.
I guess we would give the uccm otherwise permission to create & remove arbitrary RBAC resources anyway?

I would therefore prefer just giving the `system:masters` group to avoid handling RBAC in the cluster-controller

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (thz) [reply api/pkg/controller/cluster/launching.go:61]: It follows the seed metaphor quite exactly. Perhaps a bit too exactly. The seed cluster puts the seed somewhere and the user-cluster grows from it. That seed is called "cluster-seed". But I was guessing that this is slightly non-intuitive.

File: api/pkg/resources/openvpn/configmap.go (hunk 9 above)
Comment (thz) [reply api/pkg/resources/openvpn/configmap.go:98]: I can agree with the validation step, after passing the `host:port` notation. Sounds good.

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (alvaroaleman) [reply api/pkg/controller/cluster/launching.go:61]: `cluster-info-kubermatic`?

Comment (alvaroaleman) [reply api/pkg/controller/cluster/launching.go:61]: Also, please make all keys an exported constant if you want to re-use them within the user-cluster-controller-manager

File: api/pkg/controller/cluster/launching.go [hunk 26]
```diff
@@ -33,6 +33,43 @@ func (cc *Controller) clusterIsReachable(c *kubermaticv1.Cluster) (bool, error)
...
+// Creates seed ConfigMap in user-cluster
+func (cc *Controller) launchingCreateClusterSeedConfigMap(c *kubermaticv1.Cluster) error {
+	client, err := cc.userClusterConnProvider.GetClient(c)
+	if err != nil {
+		return err
+	}
+
+	name := resources.ClusterSeedConfigMapName
+	_, err = client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(name, metav1.GetOptions{})
+	if err != nil {
```
Comment (alvaroaleman) [inline api/pkg/controller/cluster/launching.go:46]: Nit: I find the code easier to read if the check for `IsNotFound` comes first, because by the time I am at the `else` below I do not know what the condition was anymore:

```
if err != nil {
  if !errors.IsNotFound(err){
    return fmt.Errorf("Error getting configmap: %v", err)
  }
  //CreateConfigMap here
}

File: api/pkg/controller/cluster/launching.go [hunk 27]
```diff
@@ -128,6 +168,17 @@ func (cc *Controller) launchingCreateClusterInfoConfigMap(c *kubermaticv1.Cluste
...
 }
 
+func (cc *Controller) getOpenVpnAddress(c *kubermaticv1.Cluster) (string, error) {
+	s, err := cc.serviceLister.Services(c.Status.NamespaceName).Get(resources.OpenVPNServerServiceName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.V(2).Infof("NotFound in getOpenVpnAddress is unexpected: %v", err)
+		}
+		return "", err
+	}
```
Comment (alvaroaleman) [inline api/pkg/controller/cluster/launching.go:180]: Please add a check for `len(s.spec.Ports == 1)` before using it

**Review by alvaroaleman — CHANGES_REQUESTED** [review]

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/controller/cluster/launching.go:61]: As customers offer a managed kubernetes hosting with kubermatic, but under a different name, `cluster-info-kubermatic` might be confusing.
And following the seed metaphor might also be confusing for endusers...

Comment (thz) [reply api/pkg/controller/cluster/launching.go:61]: Then we are back at `cluster-info-extras`. or:
*  `cluster-bootstrap-data`?
* `master-data` (as it is from the ~seed~master; customer-neutral speak)

File: api/pkg/controller/usercluster/usercluster_controller.go [hunk 28]
```diff
@@ -0,0 +1,121 @@
+package usercluster
```
Comment (alvaroaleman) [inline api/pkg/controller/usercluster/usercluster_controller.go:1]: Can we please not name this controller `usercluster` as we already have a `user-cluster-controller-manager`? Otherwise this will become the garbace can whenever someone feels something is not worth a dedicated controller and no one knows what it actually does.

Maybe just `ressourcescontroller`?

File: api/pkg/controller/usercluster/usercluster_controller.go [hunk 29]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+const (
+	// queueKey is the constant key added to the queue for deduplication.
+	queueKey = "some user-cluster object"
+)
+
+// Controller controls objects in user-cluster
+type Controller struct {
+	client          kubernetes.Interface
```
Comment (alvaroaleman) [inline api/pkg/controller/usercluster/usercluster_controller.go:29]: Can we rename this to `kubeClient` in case this controller ever manages some crd? Otherwise the name will then become ambigious

File: api/pkg/controller/usercluster/usercluster_controller.go [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+	serviceInformer k8sinformersV1.ServiceInformer) (*Controller, error) {
+
+	ucc := &Controller{
+		client:          client,
+		configMapLister: configMapInformer.Lister(),
+		serviceLister:   serviceInformer.Lister(),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "configmaps"),
+	}
+
+	configMapInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Comment (alvaroaleman) [inline api/pkg/controller/usercluster/usercluster_controller.go:47]: Doesn't this also need an informer on all resources it manages, e.G. Roles and clusterroles?

File: api/pkg/resources/data.go [hunk 31]
```diff
@@ -209,6 +211,11 @@ func (d *TemplateData) NodePortRange() string {
 	return d.nodePortRange
 }
 
+// GetMasterVpnAddress panics
+func (d *TemplateData) GetMasterVpnAddress() (string, string, error) {
+	panic("GetMasterVpnAddress not implemented for TemplateData")
```
Comment (alvaroaleman) [inline api/pkg/resources/data.go:221]: The panics will all be gone once this is part of a release?

File: api/pkg/resources/data.go (hunk 25 above)
Comment (alvaroaleman) [reply api/pkg/resources/data.go:455]: Nit: I find it easier to read to simply split by `:` then validate the `len(splittedString) == 2`. Also, the current implementation basically checks `len(splittedString) >= 1` but it must be exactly `2`

File: api/pkg/resources/resources.go (hunk 10 above)
Comment (alvaroaleman) [reply api/pkg/resources/resources.go:144]: I agree, furthermore the audit trail argument is IMHO invalid as the controller must have permissions to create more `Roles` and `RoleBindings` so indirectly it has `system:masters` anyways

File: api/cmd/user-cluster-controller-manager/main.go (hunk 15 above)
Comment (thz) [reply api/cmd/user-cluster-controller-manager/main.go]: True.

File: api/pkg/resources/data.go (hunk 25 above)
Comment (thz) [reply api/pkg/resources/data.go:455]: Actually an address `hostname` without the port should be valid and default to standard 1194 port. And then there IPv6 which has `:` in the host part. But yea - using the last `:` would break with IPv6 without port anyways... I'll rethink this.

Comment (thz) [reply api/pkg/resources/data.go:455]: ...and then in theory you can use names for ports as well. I doubt that we want to support that though.

File: api/pkg/controller/usercluster/usercluster_controller.go (hunk 28 above)
Comment (thz) [reply api/pkg/controller/usercluster/usercluster_controller.go:1]: You were the one opposing anything with `resource` in it ;-)
But wouldn't it make sense to have some distinction in the name that the controlled resources are the user/customer resources?

File: api/pkg/resources/data.go (hunk 31 above)
Comment (thz) [reply api/pkg/resources/data.go:221]: That was actually not the plan. Using a `TemplateData` in a usercluster is incorrect and also `UserClusterData` in the clustercontroller is incorrect. Those are mistakes in the code which can only be caused by seriously mixing up userclustercontroller and clustercontroller code paths. In that case I prefer to fail hard and early instead of having a no-op func to satisfy the common interface.

File: api/pkg/resources/resources.go (hunk 10 above)
Comment (thz) [reply api/pkg/resources/resources.go:144]: My audit trail argument was only about keeping the name. As henrik pointed out that is still the case when using the group. The audit trail is independent of the possibility of privilege elevation by using Roles/RoleBindings rights.

File: api/pkg/controller/cluster/launching.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/controller/cluster/launching.go:61]: `cluster-info-extras`

File: api/pkg/resources/data.go (hunk 25 above)
Comment (mrIncompetent) [reply api/pkg/resources/data.go:455]: Lets set `ip:port` as requirement. The IP can be ipv6 though.
We are the ones specifying the data, so no need to cover every edgecase

File: api/pkg/controller/usercluster/usercluster_controller.go (hunk 28 above)
Comment (alvaroaleman) [reply api/pkg/controller/usercluster/usercluster_controller.go:1]: Touchez. So let me rephrase: I think the current package name is bad because its identintical with the controller managers name but right now I do not have a good suggestion either.

**Comment by nikhita** [comment]
Closing in favour of https://github.com/kubermatic/kubermatic/pull/2671.

/close
